---
title: "RNAseq ectocervix under chronic HIV infection"
author: "Paulo Czarnewski"
date: "27 July 2023"
output:
  html_document:
    toc: true
    toc_float: true
    collapsed: false
    toc_depth: 4
    number_sections: true
    theme: flatly
    highlight: tango
    self_contained: true
    keep_md: false
editor_options: 
  chunk_output_type: console
---

# Setup

## Load Libraries

```{r libraries, message=FALSE, warning=FALSE}
library(remotes)
library(tidyverse)
library(readxl)
library(openxlsx)
remotes::install_github('czarnewski/niceRplots')

library(niceRplots)
library(edgeR)
library(EDASeq)
library(fgsea)
library(rafalib)
library(enrichR)
library(pheatmap)

# setwd("/Users/vilkal/work/Brolidens_work/Projects/chronic_HIV")
```

## Define analysis parameters

```{r analysis-parameters}
pvt <- 0.01
lft <- log2(1.5)

# Set enrichment cutoffs for sig. terms
pvalue_cutoff <- 0.1
no_genes_cutoff <- 3

```

```{r get-GEO-data, eval=FALSE}
# data paths
m_path <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE194nnn/GSE194276/suppl/GSE194276%5Fmetadata%2En%5F112%2Ecsv%2Egz"
c_path1 <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE183nnn/GSE183513/suppl/GSE183513%5FRaw%5Fgene%5Fcounts%5Fmatrix%2Ecsv%2Egz"
c_path2 <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE194nnn/GSE194276/suppl/GSE194276%5Fcounts%2Draw%2En%5F16%2Ecsv%2Egz"

download.file(m_path, destfile = "./metadata.csv.gz")
download.file(c_path1, destfile = "./Raw_gene_counts_1.csv.gz")
download.file(c_path2, destfile = "./Raw_gene_counts_2.csv.gz")

meta <- read.csv(gzfile("./metadata.csv.gz"), row.names = 1)

tmp1 <- read.csv(gzfile("./Raw_gene_counts_1.csv.gz"))
tmp2 <- read.csv(gzfile("./Raw_gene_counts_2.csv.gz"))
counts_all <- left_join(tmp1, tmp2, by=c("entrez", "symbol", "length")) %>% filter(!is.na(symbol))
write_csv(counts_all, "./Raw_gene_counts_matrix.csv")
```

## Load counts and metadata

```{r load-data}
# data paths
meta_path <- "./metadata.csv"
RNAseq_path <- "./Raw_gene_counts_matrix.csv"

counts_all <- read.csv(
  file = RNAseq_path,
  row.names = 1)

meta <- read.csv(
  file = meta_path,
  row.names = 1)

counts_all <- rowsum( counts_all[,-c(1:2) ] , counts_all$symbol )
counts_all <- counts_all[,order(colnames(counts_all))]
counts_all <- counts_all[complete.cases(counts_all), ]
dim(counts_all)

```

```{r sample-selection}
samples_ID <- c("P001", "P003", "P004", "P005", "P006", "P007", "P008", "P009", 
                "P010", "P011", "P013", "P014", "P016", "P017", "P018", "P019", 
                "P020", "P021", "P022", "P023", "P024", "P025", "P026", "P027", 
                "P028", "P029", "P031", "P032", "P033", "P037", "P038", "P039", 
                "P040", "P041", "P042", "P043", "P044", "P045", "P046", "P047", 
                "P048", "P049", "P050", "P052", "P053", "P057", "P059", "P060", 
                "P061", "P062", "P067", "P068", "P071", "P072", "P073", "P074", 
                "P075", "P079", "P080", "P081", "P087")

# n = 61 (keep 2 HIV samples on DMPA)
result_dir <- "./results/N=61/"
samples_use <- samples_ID
run <- "n=61"

# n = 59 (remove 2 HIV samples on DMPA)
# result_dir <- "../results/N=59/"
# samples_use <-  samples_ID[!(samples_ID %in% c("P010", "P032"))]
# run <- "noDMPA"

if( isFALSE(dir.exists(result_dir)) ) { dir.create(result_dir,recursive = TRUE) }
```


```{r filter-samples-to-use}
# select samples
counts <- as.matrix(counts_all[,samples_use])
meta   <- as.data.frame(meta[ samples_use ,])
dim(counts)
dim(meta)

```

***

# Differential expression

## Define model design

```{r model-design}
Group <- factor(meta$HIVstatus)
design <- model.matrix( ~ HIVstatus, data=meta )
dim(design)
```

## Fit counts with edgeR

```{r calculate-DEGs}
y <- DGEList( counts=counts, group = Group, remove.zeros=T)
y <- calcNormFactors(object = y, method = "TMM")
y <- estimateGLMCommonDisp(y,design)
y <- estimateGLMTrendedDisp(y,design)
y <- estimateGLMTagwiseDisp(y,design)
fit <- glmFit(y, design)
normcounts <- cpm(y, log = T, prior.count = 1, normalized.lib.sizes = T)
```

## Compute DGE

### HIV status

```{r define-DEGs-signifcans-level}
# Differential expression for HIV
lrt <- glmLRT(fit, coef = 2)
top <- topTags(lrt, adjust.method = "fdr", n = "all", sort.by = "p.value")[[1]] %>%
  mutate('regulation_FC' = ifelse(logFC >= lft & PValue <= pvt ,"Up", 
                               ifelse(logFC <= -lft & PValue  <= pvt, "Down", 'NotSig'))) %>%
  mutate('regulation_FDR' = ifelse(logFC >= 0 & FDR <= 0.05 ,"Up", 
                               ifelse(logFC <= -0 & FDR  <= 0.05, "Down", 'NotSig'))) %>%
  mutate('symbol' = rownames(.), .before = "logFC") 

# summary of regulation
s_df <- enframe(table(top$regulation_FDR), "Regulation", "genes")
s_df

# top 15 genes sorted by logFC
top_DEGs <- top %>%
  filter(grepl("Up|Down",.$regulation_FDR)) %>%
  group_by(regulation_FDR) %>%
  slice_max(., n = 15, order_by = abs(logFC))

knitr::kable(top_DEGs, caption = "Top 15 genes sorted by logFC")
```

```{r define-DEGs}
# NB! here you define what genes to be included in the heatmap:
top$FC <- (abs(top$logFC) > lft) & (top$PValue < pvt)
top$almost_de <- (abs(top$logFC) > lft) & (top$PValue < pvt*3)
top$fdr <- top$FDR < 0.05

```

### Plot BCV and RLE

- BCV: biological coeficient of variation
- RLE: relative log expression

```{r save-QC-plot}
pdf(paste0(result_dir,"DGE_QC_plots.pdf"),width = 8*2,height = 5,useDingbats = F)
mypar(1,2)
plotBCV(y,las=1)
plotRLE(normcounts,ylim=c(-1,1)*0.4,outline=F,las=2)
dev.off()
```

### Save normcounts and DGE results

```{r save-tables}
DEG_list <- list("DEGs list"=top, "Top 15 by logFC"=top_DEGs, "summary"=s_df)
write.xlsx( DEG_list , file = paste0(result_dir,"DGE_HIV_",run,".xlsx"))
write.csv( normcounts , file = paste0(result_dir,"normcounts.csv"), row.names = T)
write.csv( counts , file = paste0(result_dir,"raw_counts.csv"), row.names = T)

```

***

# Data exploration

## Volcano plot

### HIV status

```{r volcano-plot}
pv <- top$PValue

# top 15 by logFC and top 13 by p-value combined:
lable_symbols <- unique(c(top_DEGs$symbol, top$symbol[1:13][top$FC[1:13]] ))
top$Lable <- ifelse(grepl(paste0(lable_symbols, collapse="$|^"),top$symbol), top$symbol,NA)

pdf(paste0(result_dir,"DGE_volcano_plot_HIV_",run,".pdf"),width=5,height=5,useDingbats = F)
mypar()
plot(top$logFC,-log10(pv),xlim=c(-.9,.9)*max(abs(range(top$logFC)))*1.2,
     col=c("#00007F","grey70","firebrick")[as.numeric(as.factor(top$regulation_FDR))], pch=16,
     cex= 0.3 -log10(pv) / max(-log10(pv)) * abs(top$logFC) / max(abs(range(top$logFC)))*2 )
text(#top$logFC[top$is_top],-log10(pv)[top$is_top],
     top$logFC,-log10(pv),
     labels = top$Lable, cex=0.4,
     pos = ifelse(top$logFC[top$FC] > 0 , 2, 4),offset = 0.4 )
dev.off()
```

## PCA

### HIV status

```{r PCA-plot}
# significance: P.value <= 0.01 & Fold change => log2(1.5)
PC <- prcomp( t(normcounts[rownames(top)[top$FC],]) ,center = T,scale. = T)
pdf(paste0(result_dir,"DGE_PCA_HIV_",run,".pdf"),width = 5,height = 4.2,useDingbats = F)
par(mar=c(3,3,2,6))
plot(PC$x[,c(1,2)], pch=21, las=1, cex = 1.3,
     col = c("black","white")[as.numeric(as.factor(meta$Contraception))],
     bg = c("#F5AC10","#81858B")[as.numeric(as.factor(meta$HIVstatus))] )
legend( x = par('usr')[2], y = par('usr')[4],
        legend = levels(as.factor(meta$HIVstatus)),
        col = c("#F5AC10","#81858B"),xpd=T,pch = 16,box.lwd = 0 )
dev.off()

# significance: FDR < 0.05
PC <- prcomp( t(normcounts[rownames(top)[top$fdr],]) ,center = T,scale. = T)
pdf(paste0(result_dir,"DGE_PCA_HIV_FDR_",run,".pdf"),width = 5,height = 4.2,useDingbats = F)
par(mar=c(3,3,2,6))
plot(PC$x[,c(1,2)], pch=21, las=1, cex = 1.3,
     col = c("black","white")[as.numeric(as.factor(meta$Contraception))],
     bg = c("#F5AC10","#81858B")[as.numeric(as.factor(meta$HIVstatus))] )
legend( x = par('usr')[2], y = par('usr')[4],
        legend = levels(as.factor(meta$HIVstatus)),
        col = c("#F5AC10","#81858B"),xpd=T,pch = 16,box.lwd = 0 )
dev.off()
```

***

## Heatmap

### HIV status

```{r heatmap}
plot_heatmap.fun <- function(temp, file, filt){
  pheatmap::pheatmap(
    mat = normcounts[rownames(temp),o],
    scale = "row",
    cluster_cols = F,
    cluster_rows = F,
    labels_row = "",
    labels_col = "",
    col = colorRampPalette(c("navy","grey95","firebrick"))(99) ,
    border=F,
    clustering_method = "ward.D2",
    annotation_col = data.frame(
      age = as.numeric(meta$age),
      BV_Diagnosis_v3 = factor(meta$BV.Diagnosis),
      HIVstatus = factor(meta$HIVstatus),
      row.names = rownames(meta)) ,
    annotation_row = data.frame(
      #is_top_DE = factor(top$FC[top$FC]*1),
      abs_log2FC = abs(top$logFC)[top[[filt]]],
      log10PV = -log10(top$PValue)[top[[filt]]],
      row.names = rownames(top)[top[[filt]]]) ,
    annotation_colors = list(HIVstatus=c("neg"="#F5AC10","pos"="#81858B"),
                             BV_Diagnosis_v3=c("BV"="#F08739","Interm"="#EDB37D","Normal"="#F1E0D4")),
    filename = paste0(result_dir,file),cellwidth = 8,cellheight = 2 )
}

o <- order(meta$HIVstatus)
top$is_top <- (abs(top$logFC) > log2(2)) & (top$PValue < 0.01)
top$fdr <- top$FDR < 0.05

# significance: P.value <= 0.01 & Fold change => log2(1.5)
temp <- top[ top$FC, ]
temp <- temp[order(temp$logFC),]
plot_heatmap.fun(temp, paste0("DGE_heatmap_HIV_",run,".pdf"), filt="FC")

# significance: FDR < 0.05
temp <- top[ top$fdr, ]
temp <- temp[order(temp$logFC),]
plot_heatmap.fun(temp, paste0("DGE_heatmap_HIV_FDR_",run,".pdf"), filt="fdr")
```


***

# Functional Enrichment Analysis (EnrichR)

## Compute overrepresentation

```{r enrichment-analysis}
run_enrichment.fun <- function(h_HIV, file){
wb <- createWorkbook()

# Running enrichments for up or downregulated genes
for(i in c("HIV") ){
  message(paste0("Processing comparisson   ", i))
  cl <- get(paste0("h_",i))
  #  <- cutree(tree = get(paste0("h_",i))$tree_row, k=2)

  for(j in unique(cl)){
    message(paste0("Processing cluster   ", j))
    gene_list <- names(cl)[cl == j]

    # Gene Onthology
    a <- enrichr(genes = gene_list, databases = "GO_Biological_Process_2021")[[1]]
    a <- a[order(a$P.value,decreasing = F),]
    a <- a[a$P.value < pvalue_cutoff & as.numeric(sapply(strsplit(a[,"Overlap"],"/"),"[[", 1)) >= no_genes_cutoff,]
    a <- a[grep("regulation",a$Term,invert = T),]
    a <- separate_wider_regex(a, cols=Term, pattern=c(Term=".+(?=\\s\\()", "\\s\\(", GOid="GO\\:\\d+", "\\)"), too_few = "align_start")

    openxlsx::addWorksheet(wb, sheetName=paste0("GO_",j) )
    openxlsx::writeDataTable(wb, sheet=paste0("GO_",j), a)

    # KEGG pathways
    b <- enrichr(genes=gene_list, databases = "KEGG_2021_Human")[[1]]
    b <- b[order(b$P.value,decreasing = F),]
    b <- b[grep("hsa05",b$Term,invert = T),]   #Remove disease-related KEGG pathways ("hsa05"")
    b <- b[b$P.value < pvalue_cutoff & as.numeric(sapply(strsplit(b[,"Overlap"],"/"),"[[", 1)) >= no_genes_cutoff,]
    b$KEGGID <- sub(".*sapiens ","",b$Term); b$Term <- sub(" Homo.*","",b$Term)
    b <- data.frame(b)

    openxlsx::addWorksheet(wb, sheetName=paste0("KEGG_",j) )
    openxlsx::writeDataTable(wb, sheet=paste0("KEGG_",j), b)

    # MSigDB pathways
    c <- enrichr(genes=gene_list, databases = "MSigDB_Hallmark_2020")[[1]]
    c <- c[order(c$P.value,decreasing = F),]
    c <- c[c$P.value < pvalue_cutoff & as.numeric(sapply(strsplit(c[,"Overlap"],"/"),"[[", 1)) >= no_genes_cutoff,]

    openxlsx::addWorksheet(wb, sheetName=paste0("MSigDB_",j) )
    openxlsx::writeDataTable(wb, sheet=paste0("MSigDB_",j), c)


    # TRRUST_Transcription_Factors_2019
    d <- enrichr(genes = gene_list,
                 databases = "TRRUST_Transcription_Factors_2019")[[1]]
    d <- d[order(d$P.value,decreasing = F),]
    d <- d[grepl("human",d$Term),]
    d$Term <- sub("\ human","",d$Term)
    d <- d[d$P.value < pvalue_cutoff & as.numeric(sapply(strsplit(d[,"Overlap"],"/"),"[[", 1)) >= no_genes_cutoff,]

    openxlsx::addWorksheet(wb, sheetName=paste0("TF_TRRUST_",j) )
    openxlsx::writeDataTable(wb, sheet=paste0("TF_TRRUST_",j), d)


    # Wiki pathways
    e <- enrichr(genes=gene_list, databases = "WikiPathway_2021_Human")[[1]]
    e <- e[order(e$P.value,decreasing = F),]
    e <- e[e$P.value < pvalue_cutoff & as.numeric(sapply(strsplit(c[,"Overlap"],"/"),"[[", 1)) >= no_genes_cutoff,]
    e$wikiID <- sub(".* WP[0123456789]","WP",e$Term) ; e$Term <- sub(" WP[0123456789].*","",e$Term)

    openxlsx::addWorksheet(wb, sheetName=paste0("WikiPath_",j) )
    openxlsx::writeDataTable(wb, sheet=paste0("WikiPath_",j), e)
    
    # TRANSFAC_and_JASPAR_PWMs
    f <- enrichr(genes = gene_list,
                 databases = "TRANSFAC_and_JASPAR_PWMs")[[1]]
    f <- f[order(f$P.value,decreasing = F),]
    f <- f[f$P.value < pvalue_cutoff & as.numeric(sapply(strsplit(f[,"Overlap"],"/"),"[[", 1)) >= no_genes_cutoff,]

    openxlsx::addWorksheet(wb, sheetName=paste0("TF_TRANSFAC_",j) )
    openxlsx::writeDataTable(wb, sheet=paste0("TF_TRANSFAC_",j), f)


    # ARCHS4
    g <- enrichr(genes = gene_list,
                 databases = "ARCHS4_TFs_Coexp")[[1]]
    g <- g[order(g$P.value,decreasing = F),]
    g <- g[g$P.value < pvalue_cutoff & as.numeric(sapply(strsplit(g[,"Overlap"],"/"),"[[", 1)) >= no_genes_cutoff,]

    openxlsx::addWorksheet(wb, sheetName=paste0("TF_ARCHS4_",j) )
    openxlsx::writeDataTable(wb, sheet=paste0("TF_ARCHS4_",j), g)

  }
  saveWorkbook(wb, paste0(result_dir,file), overwrite = TRUE)
}
}

get_reg.fun <- function(FC_val, p_col, p_val){
  t <- top %>%
  mutate('reg' = ifelse(logFC >= FC_val & {{p_col}} <= p_val ,"Up", 
                        ifelse(logFC <= -FC_val & {{p_col}}  <= p_val, "Down", 'NotSig'))) %>%
    filter(!(grepl("NotSig", reg)))
  set_names(t$reg, t$symbol)
}

# significance: P.value <= 0.01 & Fold change => log2(1.5)
h_HIV <- get_reg.fun(1.5, PValue, 0.01)
run_enrichment.fun(h_HIV, paste0("GSA_",run,".xlsx"))

# significance: FDR < 0.05
h_HIV <- get_reg.fun(0, FDR, 0.05)
run_enrichment.fun(h_HIV, paste0("GSA_FDR_0.05_",run,".xlsx"))

# significance: FDR < 0.1
h_HIV <- get_reg.fun(0, FDR, 0.1)
run_enrichment.fun(h_HIV, paste0("GSA_FDR_0.1_",run,".xlsx"))

```

## Plot Functional Enrichment Analysis results

```{r enrichment-plots, results="hold", fig.width=6,fig.height=4}
#####################
# TF PPI ENRICHMENT #
#####################
# Create plot for each enrichment
n <- 10
comparison <- c("HIV")
enrichment_db <- c("GO","KEGG","TF_ARCHS4","TF_TRANSFAC","TF_TRRUST","WikiPath","MSigDB")
sheets <- readxl::excel_sheets(paste0(result_dir, "GSA_FDR_0.1_",run,".xlsx" ))

for( x in comparison ){
  pdf(paste0(result_dir,"GSA_Enrichment_barplot_",x,"_",run,".pdf"), 
      width = 6,height = 4,useDingbats = F)

for( i in enrichment_db ){

  DD <- 0
  UU <- 0
  names_DD <- ""
  names_UU <- ""
  
  f_DOWN <- read_xlsx(paste0(result_dir, "GSA_FDR_0.1_",run,".xlsx" ), sheet = paste0(i,"_Down")) %>% filter(Adjusted.P.value <= 0.05)
  f_UP <- read_xlsx(paste0(result_dir, "GSA_FDR_0.1_",run,".xlsx" ), sheet = paste0(i,"_Up")) %>% filter(Adjusted.P.value <= 0.05)
  
  if (is.data.frame(f_DOWN) && nrow(f_DOWN)>0){
    tempD_e <- f_DOWN
    tempD_e$Term <- gsub("tf ARCHS4 coexpression", "", tempD_e$Term, perl = TRUE)
    DD <- log10(tempD_e$P.value)[1:min(n,length(tempD_e$P.value))]
    names_DD <- (as.character(tempD_e$Term))[1:min(n,length(tempD_e$P.value))]
  }
  if (is.data.frame(f_UP) && nrow(f_UP)>0){
    tempU_e <- f_UP
    tempU_e$Term <- gsub("tf ARCHS4 coexpression", "", tempU_e$Term, perl = TRUE)
    UU <- rev(-log10(tempU_e$P.value)[1:min(n,length(tempU_e$P.value))])
    names_UU <- rev(as.character(tempU_e$Term)[1:min(n,length(tempU_e$P.value))])
  }
  
  max_comb <- max(abs(na.omit(c(DD,UU))))
  comb <- c( DD/max_comb , UU/max_comb )
  #comb <- c( DD , UU )
  comb <- tidyr::replace_na(comb, 0)
  if(min(comb) == 0){max_DD <- -.1}else{max_DD <- min(comb)}

  if (identical(comb, c(0,0))){next}
  
  par(mar = c(1.2, 21, 2.2, 1))  #bottom, left, top, right
  barplot( height = c(rep(NA, each=20-length(comb)), comb),  
           xlim=c(max_DD*1.5,1*1.1) , 
           cex.names = .9,
           xlab = "-log10 p-value",
           names.arg = c(rep("", each=20-length(comb)),names_DD, names_UU ), 
           horiz = T, 
           las= 1, 
           border=NA, 
           xaxt='n', #line=-.2,
           col = c(rep(NA, each=20-length(comb)), 
                   rev(colorRampPalette( c("firebrick3","gray","navy") )(19))[(comb+1)*9+1])
  )
  title(main=paste(x, stringr::str_replace_all(i, "_"," "), sep="\n"), line = 0.1, cex.main=1)
  
  points_y <- rev(seq(par("usr")[4]-1.45, by=-1.2, length=length(c(UU,DD))))
           points( comb , points_y,
                   bg=colorRampPalette( c("navy","gray","firebrick3") )(19)[(comb+1)*9+1],
                   pch=21, col="black",cex=1.5)
           lines(x=c(0, 0), y= c(points_y[1]-.6, par("usr")[4]-1)) 
           lines(x=c(-2/max_comb,-2/max_comb),
                 y= c(points_y[1]-.6, par("usr")[4]-1), lty=c(3)) # pval=0.01
           lines(x=c(2/max_comb,2/max_comb),
                 y= c(points_y[1]-.6, par("usr")[4]-1), lty=c(3)) # pval=0.01
           text(0,points_y[1]-1.5, "-log10 p-value", xpd=T)
    }
  dev.off()
  }
```

The end.
:wq
:)